### ISSUE TITLE
[no-non-null-asserted-optional-chain] false negative for `foo1?.bar!.baz`

### ISSUE THREAD

**Issue by xcatliu** [typescript-eslint/typescript-eslint#2412]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/no-non-null-asserted-optional-chain": "error"
  }
}
```

```TS
let foo: { bar: any } | undefined;

foo?.bar!.baz;
```

**Expected Result**

As [the docs](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/no-non-null-asserted-optional-chain.md) described, it should report an error.

**Actual Result**

No error.

![image](https://user-images.githubusercontent.com/5453359/90853026-a11a7580-e3ab-11ea-88e4-f0172204fecd.png)

**Additional Info**

`foo?.bar!;` report an error as expected.

![image](https://user-images.githubusercontent.com/5453359/90853068-b0012800-e3ab-11ea-8325-741596640a48.png)
![image](https://user-images.githubusercontent.com/5453359/90853090-be4f4400-e3ab-11ea-9c1a-32e2ae9d89ec.png)

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `3.9.1` |
| `@typescript-eslint/parser`        | `3.9.1` |
| `TypeScript`                       | `4.0.2` |
| `ESLint`                           | `7.7.0` |
| `node`                             | `12.16.1` |

**Comment by bradzacher** [comment]
The docs are out of date. TS3.9 changed how optional chaining works.

Previously `foo?.bar!.baz` meant `(foo?.bar).baz` - which means "assume the result of `foo?.bar` is non-null.
As of TS3.9, `foo?.bar!.baz` now just applies the non-null assertion to the property itself.

So it's no longer bad practice to do this.

It is however still bad practice to do `foo?.bar.baz!`, because that applies the non-null assertion to the _entire_ chain.
